### PR TITLE
Add java compiler error reporting release note

### PR DESCRIPTION
**Pull request by hegyibalint**
### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

File: platforms/documentation/docs/src/docs/release/notes.md
Comment (donat) [inline platforms/documentation/docs/src/docs/release/notes.md:53]: I think we should start with the context. What's the problem we wanted to solve:
- in previous Gradle versions, the build solely relied on the compiler output
- If build failed, the message just had an advice to find it in the build log
- For large builds and for CI this means tedious scrolling finding the corresponding compile task output

File: platforms/documentation/docs/src/docs/release/notes.md
Comment (donat) [inline platforms/documentation/docs/src/docs/release/notes.md:85]: It's not really a limitation, it's the scope of the feature. I propose we soften the sentence a little bit, by saying
```suggestion
Note, that the output remains unchanged when the Java compilation task finishes with warnings only. 
```
Also, this should be the last sentence of the release notes.

File: platforms/documentation/docs/src/docs/release/notes.md
Comment (donat) [inline platforms/documentation/docs/src/docs/release/notes.md:86]: This belongs to the roadmap.

```suggestion
```

File: platforms/documentation/docs/src/docs/release/notes.md
Comment (donat) [inline platforms/documentation/docs/src/docs/release/notes.md:75]: I'm not sure if we discussed this deeply; but I don't think we should have this line in the output. The ID (`compilation:java:java-compilation-warning`) is an internal detail for the problems API and the display name (`Java compilation warning`) contains information already present in the next line.

    sample-project/src/main/java/Problem1.java:6: warning: [cast] redundant cast to String
                                                  ^

File: platforms/documentation/docs/src/docs/release/notes.md
Comment (donat) [inline platforms/documentation/docs/src/docs/release/notes.md:87]: I'm not sure on this but we might want to reiterate that this feature is built on top of the Problems API. The important bit for me is that we expose structural information consistently to many consumers: build scans, IDEs, and the console.  We could also add a link to the 8.9 release notes where we first mentioned the compiler integration.

**Review by donat — CHANGES_REQUESTED** [review]
A few items to discuss.

File: platforms/documentation/docs/src/docs/release/notes.md
Comment (hegyibalint) [reply platforms/documentation/docs/src/docs/release/notes.md:53]: FYI, I've wanted to tell this story, but I was unsure how much storytelling we want to do in the release notes. I guess by your comment, you also think it would be worthwhile, so I will expand upon the details

File: platforms/documentation/docs/src/docs/release/notes.md
Comment (hegyibalint) [reply platforms/documentation/docs/src/docs/release/notes.md:75]: I understand your concern, but I don't find the discussion about this on the PR, and you didn't block the PR neither.

I can still fix this, just noting that this misunderstanding could've been avoided.

File: platforms/documentation/docs/src/docs/release/notes.md
Comment (hegyibalint) [reply platforms/documentation/docs/src/docs/release/notes.md:86]: How can people find the roadmap?

Comment (donat) [reply platforms/documentation/docs/src/docs/release/notes.md:86]: https://blog.gradle.org/roadmap-announcement

File: platforms/documentation/docs/src/docs/release/notes.md
Comment (hegyibalint) [reply platforms/documentation/docs/src/docs/release/notes.md:87]: I don't think this is relevant here for the users reading the release notes, especially as we already had the integration enabled in 8.9.

This is our internal adoption of using the data in a new format.

File: platforms/documentation/docs/src/docs/release/notes.md
Comment (hegyibalint) [reply platforms/documentation/docs/src/docs/release/notes.md:85]: I think the solution is somewhere in the middle: we should denote that if there is no failure, there are no warnings displayed.

File: platforms/documentation/docs/src/docs/release/notes.md
Comment (hegyibalint) [reply platforms/documentation/docs/src/docs/release/notes.md:86]: Also, I want the users to know that we still care about reporting warnings when only warnings happen, but it is not ready yet. If you think from the user perspective, this story is not yet complete, and I want to emphasize that further features are coming.

Comment (hegyibalint) [reply platforms/documentation/docs/src/docs/release/notes.md:86]: > https://blog.gradle.org/roadmap-announcement

Maybe I haven't asked the correct question: how the users should know that this is part of a roadmap if we don't include the roadmap.
Do yo want me to include it?

Comment (hegyibalint) [reply platforms/documentation/docs/src/docs/release/notes.md:86]: Am I wrong not to see a roadmap item for Problems API reporting on the roadmap? I can see the Java one, but I don't see the universal reporting features.

File: platforms/documentation/docs/src/docs/release/notes.md
Comment (donat) [reply platforms/documentation/docs/src/docs/release/notes.md:53]: I've added @lkasso to this review explicitly to help us with the content.

File: platforms/documentation/docs/src/docs/release/notes.md
Comment (donat) [reply platforms/documentation/docs/src/docs/release/notes.md:86]: So my comment was about the fact that we should talk about stuff that we have in the release notes and not promise anything. Unlike the published release notes, we can adjust the roadmap later.

Comment (hegyibalint) [reply platforms/documentation/docs/src/docs/release/notes.md:86]: Might not have enough time to make PR

Comment (donat) [reply platforms/documentation/docs/src/docs/release/notes.md:86]: Maybe we should only mention that we'll have more enhancements as part of the [`Clean and actionable error reporting`](https://github.com/gradle/build-tool-roadmap/issues/49) roadmap item. 

Also, I'm happy to adjust the description on the roadmap based on your feedback.

File: platforms/documentation/docs/src/docs/release/notes.md
Comment (donat) [reply platforms/documentation/docs/src/docs/release/notes.md:87]: Ok.

File: platforms/documentation/docs/src/docs/release/notes.md
Comment (hegyibalint) [reply platforms/documentation/docs/src/docs/release/notes.md:86]: That being said, the solution uses a common rendering class: https://github.com/gradle/gradle/blob/282f62db0ac5e4f59c442229bafcb7feb0ff4593/platforms/ide/problems-rendering/src/main/java/org/gradle/problems/internal/rendering/ProblemRenderer.java#L32

I didn't make any strategy pattern here on differentiating the rendering, as I believed we agreed on an uniform look between a (coming) problem console reporting, and this report

Comment (hegyibalint) [reply platforms/documentation/docs/src/docs/release/notes.md]: Pushed the new test, PTAL

File: platforms/documentation/docs/src/docs/release/notes.md
Comment (hegyibalint) [reply platforms/documentation/docs/src/docs/release/notes.md]: Pushed the new text, then resolve or comment

Comment (hegyibalint) [reply platforms/documentation/docs/src/docs/release/notes.md]: Left CI story out, my bad

Comment (hegyibalint) [reply platforms/documentation/docs/src/docs/release/notes.md]: Though new text might be enough, and we don't need to shoehorn in CI there.

File: platforms/documentation/docs/src/docs/release/notes.md
Comment (hegyibalint) [reply platforms/documentation/docs/src/docs/release/notes.md]: Pushed the new text, then resolve or comment

**Comment by lkasso** [comment]
@donat @hegyibalint Is this supposed to go into 8.10 release notes or 8.11?

File: platforms/documentation/docs/src/docs/release/notes.md
Comment (lkasso) [inline platforms/documentation/docs/src/docs/release/notes.md:53]:
```suggestion
In previous Gradle versions, identifying the cause of a failed compilation was often cumbersome. 
Users had to scroll through lengthy build logs to find the task output containing the compiler error.
```

File: platforms/documentation/docs/src/docs/release/notes.md
Comment (lkasso) [inline platforms/documentation/docs/src/docs/release/notes.md:55]:
```suggestion
This process was particularly challenging when using the `--continue` flag, as handling multiple failures could make it difficult to pinpoint the exact cause of an error.
```

File: platforms/documentation/docs/src/docs/release/notes.md
Comment (lkasso) [inline platforms/documentation/docs/src/docs/release/notes.md:57]:
```suggestion
Gradle now has the ability to collect and report per-task failures in the "What went wrong" section at the bottom of the build log.
```

File: platforms/documentation/docs/src/docs/release/notes.md
Comment (lkasso) [inline platforms/documentation/docs/src/docs/release/notes.md:58]:
```suggestion
This enhancement supplements the usual task output and provides a more straightforward way to identify issues.
```

File: platforms/documentation/docs/src/docs/release/notes.md
Comment (lkasso) [inline platforms/documentation/docs/src/docs/release/notes.md:60]:
```suggestion
For example, a simple failure now appears in the bottom of the log as:
```

File: platforms/documentation/docs/src/docs/release/notes.md
Comment (lkasso) [inline platforms/documentation/docs/src/docs/release/notes.md:72]:
```suggestion
Warnings during compilation are also included in the report:

```

File: platforms/documentation/docs/src/docs/release/notes.md
Comment (lkasso) [inline platforms/documentation/docs/src/docs/release/notes.md:88]:
```suggestion
Note that this report only appears when there are _failures_. If only warnings occur (and `-Werror` is not set), this report will not be shown.
```

File: platforms/documentation/docs/src/docs/release/notes.md
Comment (lkasso) [inline platforms/documentation/docs/src/docs/release/notes.md:90]:
```suggestion
This feature also works with the [`--continue`](https://docs.gradle.org/current/userguide/command_line_interface.html#sec:continue_build_on_failure) flag, providing a per-task report of all compilation failures at the bottom of the log.
```

**Review by lkasso — CHANGES_REQUESTED** [review]
I left some updates on the wording.
Is this supposed to go into 8.10?

File: platforms/documentation/docs/src/docs/release/notes.md
Comment (hegyibalint) [reply platforms/documentation/docs/src/docs/release/notes.md:55]: I think we lost some meaning there:
 - The key point is that this feature helps with long logs
 - I've tried to convey that if for example you have long logs, and use `--continue` (tbh, not a super commonly used feature), you might have to spend a decent time on finding the root issue of your compilation

File: platforms/documentation/docs/src/docs/release/notes.md
Comment (hegyibalint) [reply platforms/documentation/docs/src/docs/release/notes.md:88]: Opps, good catch with the _not_! 🎉

**Comment by hegyibalint** [comment]
> I left some updates on the wording.
> Is this supposed to go into 8.10?

It was my bad, rebased to 8.10

**Review by donat — COMMENTED** [review]
The content is ok from my point of view. I have one more items to add here, otherwise I leave it to Laura to 👍 the changes.

**Review by lkasso — APPROVED** [review]

**Comment by lkasso** [comment]
I left two more comments, once resolved, the PR LGTM.

File: platforms/documentation/docs/src/docs/release/notes.md
Comment (big-guy) [inline platforms/documentation/docs/src/docs/release/notes.md:96]: This should not use an absolute URL.  It needs to be relative like:
```suggestion
The feature also works with the [`--continue`](userguide/command_line_interface.html#sec:continue_build_on_failure) flag, and the bottom report will contain a per-task report of all the compilation failures.
```

**Review by big-guy — CHANGES_REQUESTED** [review]

**Comment by hegyibalint** [comment]
Thank you all for the feedback!

**Pull request merged**
